db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/ValueNode.java

```diff
@@ -1622,6 +1624,46 @@ public boolean collectExpressions(Map<Integer, Set<ValueNode>> exprMap) {
         return collectSingleExpression(exprMap);
     }
 
+    /*
+     * Add a cast node to the rightOperand if it's of string type and need to be compared
+     * to leftOperand
+     * If we are comparing a non-string with a string type, then we
+     * must prevent the non-string value from being used to probe into
+     * an index on a string column. This is because the string types
+     * are all of low precedence, so the comparison rules of the non-string
+     * value are used, so it may not find values in a string index because
+     * it will be in the wrong order. So, cast the string value to its
+     * own type. This is easier than casting it to the non-string type,
+     * because we would have to figure out the right length to cast it to.
+     * @return the new rightOperand
+     */
+    protected ValueNode addCastNodeForStringToNonStringComparison(
+            ValueNode leftOperand, ValueNode rightOperand) throws StandardException {
```

Comment:
Does it make sense to put this function in `BinaryOperatorNode`? It seems not used in any other context.

Reply:
It is used in both `BinaryComparisonOperatorNode` and `BinaryListOperatorNode` whose closest common ancestor is `ValueNode`